src/python/WMCore/MicroService/Unified/Common.py

```diff
@@ -101,10 +128,43 @@ def phedexValidBlocks(datasets, phedexUrl):
         rows = json.loads(row['data'])
         dsetBlockSize.setdefault(dataset, {})  # flat dict in the format of blockName: blockSize
         for item in rows['phedex']['block']:
-            dsetBlockSize[dataset].update({item['name']: item['bytes']})
+            block = {item['name']: {'blockSize': item['bytes'], 'locations': []}}
+            for repli in item['replica']:
+                if repli['complete'] == 'y' and repli['subscribed'] == 'y' and repli['group'] == group:
+                    block[item['name']]['locations'].append(repli['node'])
+            dsetBlockSize[dataset].update(block)
     return dsetBlockSize
 
 
+def getPileupSubscriptions(datasets, phedexUrl, group="DataOps", percentMin=99):
+    """
+    Provided a list of datasets, find dataset level subscriptions where it's
+    as complete as `percent_min`.
+    :param datasets: list of dataset names
+    :param phedexUrl: a string with the PhEDEx URL
+    :param group: PhEDEx group
+    :param percent_min: only return subscriptions that are this complete
+    :return: a dictionary of datasets and a list of their location
+    """
+    #FIXME: implement the same logic for Rucio
+    locationByDset = {}
+    if not datasets:
+        return locationByDset
+
+    url = "%s/subscriptions?group=%s&percent_min=%s&dataset=%s"
+    urls = [url % (phedexUrl, group, percentMin, dset) for dset in datasets]
+    data = multi_getdata(urls, ckey(), cert())
+
+    for row in data:
+        dataset = row['url'].rsplit('=')[-1]
+        rows = json.loads(row['data'])
```

Comment:
again there is no error conditioning checks here. And part below is phedex specific.

Reply:
I applied those changes to all functions in this module